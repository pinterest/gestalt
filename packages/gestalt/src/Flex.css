/*
This file was generated by running

  $ ./scripts/flex.js > packages/gestalt/src/Flex.css

*/

.Flex {
  display: flex;
}

.FlexItem {
  display: block;
}

.columnGap0 {
  margin-bottom: 0;
  margin-top: 0;
}

.columnGap0 > .FlexItem {
  margin-bottom: 0;
  margin-top: 0;
}

.rowGap0 {
  margin-left: 0;
  margin-right: 0;
}

.rowGap0 > .FlexItem {
  margin-left: 0;
  margin-right: 0;
}

.columnGap1 {
  margin-bottom: -2px;
  margin-top: -2px;
}

.columnGap1 > .FlexItem {
  margin-bottom: 2px;
  margin-top: 2px;
}

.rowGap1 {
  margin-left: -2px;
  margin-right: -2px;
}

.rowGap1 > .FlexItem {
  margin-left: 2px;
  margin-right: 2px;
}

.columnGap2 {
  margin-bottom: -4px;
  margin-top: -4px;
}

.columnGap2 > .FlexItem {
  margin-bottom: 4px;
  margin-top: 4px;
}

.rowGap2 {
  margin-left: -4px;
  margin-right: -4px;
}

.rowGap2 > .FlexItem {
  margin-left: 4px;
  margin-right: 4px;
}

.columnGap3 {
  margin-bottom: -6px;
  margin-top: -6px;
}

.columnGap3 > .FlexItem {
  margin-bottom: 6px;
  margin-top: 6px;
}

.rowGap3 {
  margin-left: -6px;
  margin-right: -6px;
}

.rowGap3 > .FlexItem {
  margin-left: 6px;
  margin-right: 6px;
}

.columnGap4 {
  margin-bottom: -8px;
  margin-top: -8px;
}

.columnGap4 > .FlexItem {
  margin-bottom: 8px;
  margin-top: 8px;
}

.rowGap4 {
  margin-left: -8px;
  margin-right: -8px;
}

.rowGap4 > .FlexItem {
  margin-left: 8px;
  margin-right: 8px;
}

.columnGap5 {
  margin-bottom: -10px;
  margin-top: -10px;
}

.columnGap5 > .FlexItem {
  margin-bottom: 10px;
  margin-top: 10px;
}

.rowGap5 {
  margin-left: -10px;
  margin-right: -10px;
}

.rowGap5 > .FlexItem {
  margin-left: 10px;
  margin-right: 10px;
}

.columnGap6 {
  margin-bottom: -12px;
  margin-top: -12px;
}

.columnGap6 > .FlexItem {
  margin-bottom: 12px;
  margin-top: 12px;
}

.rowGap6 {
  margin-left: -12px;
  margin-right: -12px;
}

.rowGap6 > .FlexItem {
  margin-left: 12px;
  margin-right: 12px;
}

.columnGap7 {
  margin-bottom: -14px;
  margin-top: -14px;
}

.columnGap7 > .FlexItem {
  margin-bottom: 14px;
  margin-top: 14px;
}

.rowGap7 {
  margin-left: -14px;
  margin-right: -14px;
}

.rowGap7 > .FlexItem {
  margin-left: 14px;
  margin-right: 14px;
}

.columnGap8 {
  margin-bottom: -16px;
  margin-top: -16px;
}

.columnGap8 > .FlexItem {
  margin-bottom: 16px;
  margin-top: 16px;
}

.rowGap8 {
  margin-left: -16px;
  margin-right: -16px;
}

.rowGap8 > .FlexItem {
  margin-left: 16px;
  margin-right: 16px;
}

.columnGap9 {
  margin-bottom: -18px;
  margin-top: -18px;
}

.columnGap9 > .FlexItem {
  margin-bottom: 18px;
  margin-top: 18px;
}

.rowGap9 {
  margin-left: -18px;
  margin-right: -18px;
}

.rowGap9 > .FlexItem {
  margin-left: 18px;
  margin-right: 18px;
}

.columnGap10 {
  margin-bottom: -20px;
  margin-top: -20px;
}

.columnGap10 > .FlexItem {
  margin-bottom: 20px;
  margin-top: 20px;
}

.rowGap10 {
  margin-left: -20px;
  margin-right: -20px;
}

.rowGap10 > .FlexItem {
  margin-left: 20px;
  margin-right: 20px;
}

.columnGap11 {
  margin-bottom: -22px;
  margin-top: -22px;
}

.columnGap11 > .FlexItem {
  margin-bottom: 22px;
  margin-top: 22px;
}

.rowGap11 {
  margin-left: -22px;
  margin-right: -22px;
}

.rowGap11 > .FlexItem {
  margin-left: 22px;
  margin-right: 22px;
}

.columnGap12 {
  margin-bottom: -24px;
  margin-top: -24px;
}

.columnGap12 > .FlexItem {
  margin-bottom: 24px;
  margin-top: 24px;
}

.rowGap12 {
  margin-left: -24px;
  margin-right: -24px;
}

.rowGap12 > .FlexItem {
  margin-left: 24px;
  margin-right: 24px;
}
