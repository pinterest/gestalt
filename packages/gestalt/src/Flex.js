// @flow strict
import { Children, type Node } from 'react';
import FlexItem from './FlexItem.js';
import styles from './Flex.css';
import wrapWithComponent from './utils/wrapWithComponent.js';
import { buildStyles } from './boxTransforms.js';

type Dimension = number | string;

type Props = {|
  /**
   * Aligns the flex container's lines within when there is extra space in the cross axis, similar to how `justify-content` aligns individual items within the main axis.
   */
  alignContent?: 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly' | 'stretch',
  /**
   * Defines the default behaviour for how flex items are laid out along the cross axis on the current line. Think of it as the `justify-content` version for the cross axis (perpendicular to the main axis).
   */
  alignItems?: 'start' | 'end' | 'center' | 'baseline' | 'stretch',
  /**
   * Allows the default alignment (or the one specified by `align-items`) to be overridden for individual flex items.
   */
  alignSelf?: 'auto' | 'start' | 'end' | 'center' | 'baseline' | 'stretch',
  /**
   *
   */
  children?: Node,
  /**
   * Establishes the main axis, thus defining the direction flex items are placed in the flex container.
   */
  direction?: 'row' | 'column',
  /**
   * Defines how a flex item will be sized. "grow", equivalent to "flex: 1 1 auto", will size Flex relative to its parent, growing and shrinking based on available space. "shrink", equivalent to "flex: 0 1 auto" (the browser default), allows Flex to shrink if compressed but not grow if given extra space. Finally, "none", equivalent to "flex: 0 0 auto", preserves Flex's size based on child content regardless of its container's size.
   * Default: 'shrink'
   */
  flex?: 'grow' | 'shrink' | 'none',
  /**
   * Defines spacing between each child along the main axis.
   */
  gap?: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12,
  /**
   * Use numbers for pixels: `height={100}` and strings for percentages: `height="100%"`.
   */
  height?: Dimension,
  /**
   * Defines the alignment along the main axis. It helps distribute extra free space left over when either all the flex items on a line are inflexible, or are flexible but have reached their maximum size. It also exerts some control over the alignment of items when they overflow the line.
   */
  justifyContent?: 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly',
  /**
   * Use numbers for pixels: `maxHeight={100}` and strings for percentages: `maxHeight="100%"`.
   */
  maxHeight?: Dimension,
  /**
   * Use numbers for pixels: `maxWidth={100}` and strings for percentages: `maxWidth="100%"`.
   */
  maxWidth?: Dimension,
  /**
   * Use numbers for pixels: `minHeight={100}` and strings for percentages: `minHeight="100%"`.
   */
  minHeight?: Dimension,
  /**
   * Use numbers for pixels: `minWidth={100}` and strings for percentages: `minWidth="100%"`.
   */
  minWidth?: Dimension,
  /**
   * Defines how to handle content that extends beyond the Flex container.
   */
  overflow?: 'visible' | 'hidden' | 'scroll' | 'scrollX' | 'scrollY' | 'auto',
  /**
   * Use numbers for pixels: `width={100}` and strings for percentages: `width="100%"`.
   */
  width?: Dimension,
  /**
   * By default, flex items will all try to fit onto one line. You can change that and allow the items to wrap onto multiple lines, from top to bottom.
   */
  wrap?: boolean,
|};

const allowedProps = [
  'alignContent',
  'alignItems',
  'alignSelf',
  'children',
  'direction',
  'flex',
  'height',
  'justifyContent',
  'maxHeight',
  'maxWidth',
  'minHeight',
  'minWidth',
  'overflow',
  'width',
  'wrap',
];

/**
 * [Flex](https://gestalt.pinterest.systems/flex) is a layout component with a very limited subset of the props available to [Box](https://gestalt.pinterest.systems/box) and a special prop of its own.

 * Use this component for flexbox layouts, especially when even spacing between elements is desired (see the `gap` prop!).
 *
 * ![Flex light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Flex.spec.mjs-snapshots/Flex-chromium-darwin.png)
 *
 */
export default function Flex({
  alignItems,
  children: childrenProp,
  direction = 'row',
  gap = 0,
  justifyContent,
  ...rest
}: Props): Node {
  const children = gap
    ? Children.map(childrenProp, (child, index) => {
        if (child === null || child === undefined) {
          return null;
        }
        return wrapWithComponent({
          element: child,
          Component: FlexItem,
          props: {
            key: index,
          },
        });
      }).filter(Boolean)
    : childrenProp;

  const { passthroughProps, propsStyles } = buildStyles<Props>({
    baseStyles: `${styles.Flex} ${styles[`${direction}Gap${gap}`]}`,
    props: {
      ...rest,
      alignItems,
      children,
      direction,
      justifyContent,
    },
    allowlistProps: allowedProps,
  });

  return <div {...passthroughProps} {...propsStyles} />;
}

Flex.Item = FlexItem;

Flex.displayName = 'Flex';
