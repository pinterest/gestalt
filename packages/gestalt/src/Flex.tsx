import { Children, ReactNode } from 'react';
import { buildStyles } from './boxTransforms';
import styles from './Flex.css';
import FlexItem from './FlexItem';
import wrapWithComponent from './utils/wrapWithComponent';

type Dimension = number | string;
type Gap = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16;

type Props = {
  /**
   * Aligns the flex container's lines within when there is extra space in the cross axis, similar to how `justify-content` aligns individual items within the main axis.
   */
  alignContent?: 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly' | 'stretch';
  /**
   * Defines the default behaviour for how flex items are laid out along the cross axis on the current line. Think of it as the `justify-content` version for the cross axis (perpendicular to the main axis).
   *
   * Also available in responsive sizes: `smAlignItems`, `mdAlignItems`, `lgAlignItems`. See the [Screen size page](https://gestalt.pinterest.systems/foundations/screen_sizes#Web-(px)) to learn more about viewport breakpoints in web.
   */
  alignItems?: 'start' | 'end' | 'center' | 'baseline' | 'stretch';
  /**
   * The alignItems style in sm or larger viewports. See the [Screen size page](https://gestalt.pinterest.systems/foundations/screen_sizes#Web-(px)) to learn more about viewport breakpoints in web.
   */
  smAlignItems?: 'start' | 'end' | 'center' | 'baseline' | 'stretch';
  /**
   * The alignItems style in md or larger viewports. See the [Screen size page](https://gestalt.pinterest.systems/foundations/screen_sizes#Web-(px)) to learn more about viewport breakpoints in web.
   */
  mdAlignItems?: 'start' | 'end' | 'center' | 'baseline' | 'stretch';
  /**
   * The alignItems style in lg or larger viewports. See the [Screen size page](https://gestalt.pinterest.systems/foundations/screen_sizes#Web-(px)) to learn more about viewport breakpoints in web.
   */
  lgAlignItems?: 'start' | 'end' | 'center' | 'baseline' | 'stretch';
  /**
   * Allows the default alignment (or the one specified by `align-items`) to be overridden for individual flex items.
   */
  alignSelf?: 'auto' | 'start' | 'end' | 'center' | 'baseline' | 'stretch';
  /**
   * Note that each child will be automatically wrapped in [Flex.Item](https://gestalt.pinterest.systems/web/flex#Flex.Item) to apply various styles. If specific flex styles are needed on a child, you can manually wrap it in Flex.Item to apply those styles. See [the Applying flex properties to children example](https://gestalt.pinterest.systems/web/flex#Applying-flex-properties-to-children) to learn more.
   */
  children?: ReactNode;
  /**
   * Used to identify the element for testing purposes.
   */
  dataTestId?: string;
  /**
   * Establishes the main axis, thus defining the direction flex items are placed in the flex container.
   */
  direction?: 'row' | 'column';
  /**
   * Defines how a flex item will be sized. "grow", equivalent to "flex: 1 1 auto", will size Flex relative to its parent, growing and shrinking based on available space. "shrink", equivalent to "flex: 0 1 auto" (the browser default), allows Flex to shrink if compressed but not grow if given extra space. Finally, "none", equivalent to "flex: 0 0 auto", preserves Flex's size based on child content regardless of its container's size.
   * Default: 'shrink'
   */
  flex?: 'grow' | 'shrink' | 'none';
  /**
   * Defines spacing between each child along the main and cross axes. Use an object to define different spacing for rows and columns. See the [Gap](https://gestalt.pinterest.systems/web/flex#Gap) variant to learn more.
   */
  gap?:
    | Gap
    | {
        row: Gap;
        column: Gap;
      };
  /**
   * Use numbers for pixels: `height={100}` and strings for percentages: `height="100%"`.
   */
  height?: Dimension;
  /**
   * Defines the alignment along the main axis. It helps distribute extra free space left over when either all the flex items on a line are inflexible, or are flexible but have reached their maximum size. It also exerts some control over the alignment of items when they overflow the line.
   */
  justifyContent?: 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly';
  /**
   * Use numbers for pixels: `maxHeight={100}` and strings for percentages: `maxHeight="100%"`.
   */
  maxHeight?: Dimension;
  /**
   * Use numbers for pixels: `maxWidth={100}` and strings for percentages: `maxWidth="100%"`.
   */
  maxWidth?: Dimension;
  /**
   * Use numbers for pixels: `minHeight={100}` and strings for percentages: `minHeight="100%"`.
   */
  minHeight?: Dimension;
  /**
   * Use numbers for pixels: `minWidth={100}` and strings for percentages: `minWidth="100%"`.
   */
  minWidth?: Dimension;
  /**
   * Defines how to handle content that extends beyond the Flex container.
   */
  overflow?: 'visible' | 'hidden' | 'scroll' | 'scrollX' | 'scrollY' | 'auto';
  /**
   * Use numbers for pixels: `width={100}` and strings for percentages: `width="100%"`.
   */
  width?: Dimension;
  /**
   * By default, flex items will all try to fit onto one line. You can change that and allow the items to wrap onto multiple lines, from top to bottom.
   */
  wrap?: boolean;
};

const allowedProps = [
  'alignContent',
  'alignItems',
  'smAlignItems',
  'mdAlignItems',
  'lgAlignItems',
  'alignSelf',
  'children',
  'direction',
  'flex',
  'height',
  'justifyContent',
  'maxHeight',
  'maxWidth',
  'minHeight',
  'minWidth',
  'overflow',
  'width',
  'wrap',
];

/**
 * [Flex](https://gestalt.pinterest.systems/web/flex) is a layout component with a very limited subset of the props available to [Box](https://gestalt.pinterest.systems/web/box) and a special prop of its own.

 * Use this component for flexbox layouts, especially when even spacing between elements is desired (see the `gap` prop!).
 *
 * ![Flex light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Flex.spec.ts-snapshots/Flex-chromium-darwin.png)
 *
 */
export default function Flex({
  children: childrenProp,
  dataTestId,
  direction = 'row',
  gap = 0,
  justifyContent,
  ...rest
}: Props) {
  const children = gap
    ? // @ts-expect-error - TS2533 - Object is possibly 'null' or 'undefined'.
      Children.map(childrenProp, (child, index) => {
        if (child === null || child === undefined) {
          return null;
        }
        return wrapWithComponent({
          element: child,
          Component: FlexItem,
          props: {
            // @ts-expect-error - TS2322 - Type '{ key: number; }' is not assignable to type 'Props'.
            key: index,
          },
        });
      }).filter(Boolean)
    : childrenProp;

  const gapStyles = `${styles[`rowGap${typeof gap === 'number' ? gap : gap.row}`]} ${
    styles[`columnGap${typeof gap === 'number' ? gap : gap.column}`]
  }`;

  const { passthroughProps, propsStyles } = buildStyles<Props>({
    baseStyles: `${styles.Flex} ${gapStyles}`,
    props: {
      ...rest,
      children,
      direction,
      justifyContent,
    },
    allowlistProps: allowedProps,
  });

  // @ts-expect-error - TS2322 - Type '{ "data-test-id": string | undefined; className: string | null | undefined; style: InlineStyle | null | undefined; alignContent?: "center" | "start" | "end" | "stretch" | "between" | "around" | "evenly" | undefined; ... 18 more ...; wrap?: boolean | undefined; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
  return <div {...passthroughProps} {...propsStyles} data-test-id={dataTestId} />;
}

Flex.Item = FlexItem;

Flex.displayName = 'Flex';
