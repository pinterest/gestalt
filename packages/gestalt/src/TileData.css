.baseTile {
  composes: flex from "./Layout.css";
  composes: paddingX4 from "./boxWhitespace.css";
  composes: paddingY4 from "./boxWhitespace.css";
  composes: rounding4 from "./Borders.css";
  composes: borderBox from "./Layout.css";
  composes: sizeLg from "./Borders.css";
  composes: solid from "./Borders.css";
  composes: transparentBorder from "./Borders.css";
  box-sizing: border-box;
  width: min-content;
}

.tileWidth {
  max-width: 196px;
}

@media (--g-sm) {
  .tileWidth {
    max-width: none;
  }
}

.hovered {
  composes: pointer from "./Cursor.css";
  composes: secondary from "./Colors.css";
}

.focused {
  composes: accessibilityOutlineFocus from "./Focus.css";
}

.transparent {
  composes: transparentBg from "./Colors.css";
}

.selected {
  composes: secondary from "./Colors.css";
}

.disabled {
  composes: lightGrayBg from "./Colors.css";
  composes: transparentBorder from "./Borders.css";
  cursor: default;
  pointer-events: none;
}
