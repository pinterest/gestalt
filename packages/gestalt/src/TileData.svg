<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><path fill="#005C62" d=""/></svg>
