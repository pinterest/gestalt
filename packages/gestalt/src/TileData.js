// @flow strict
import { type Node as ReactNode, useId } from 'react';
import classnames from 'classnames';
import Box from './Box';
import InternalCheckbox from './Checkbox/InternalCheckbox';
import { useColorScheme } from './contexts/ColorSchemeProvider';
import InternalDatapoint from './Datapoint/InternalDatapoint';
import Flex from './Flex';
import TapArea from './TapArea';
import styles from './TileData.css';
import getCheckboxColors from './utils/datavizcolors/getCheckboxColor';
import getDataVisualizationColor from './utils/datavizcolors/getDataVisualizationColor';
import MaybeTooltip from './utils/MaybeTooltip';
import useInteractiveStates from './utils/useInteractiveStates';
import { type Indexable } from './zIndex';

type TooltipProps = {
  accessibilityLabel?: string,
  inline?: boolean,
  idealDirection?: 'up' | 'right' | 'down' | 'left',
  text: string | $ReadOnlyArray<string>,
  zIndex?: Indexable,
};

export type TileChangeHandler = ({
  event:
    | SyntheticMouseEvent<HTMLDivElement>
    | SyntheticKeyboardEvent<HTMLDivElement>
    | SyntheticMouseEvent<HTMLAnchorElement>
    | SyntheticKeyboardEvent<HTMLAnchorElement>,
  selected: boolean,
  id?: string,
}) => void;

export type DataVisualizationColors =
  | '01'
  | '02'
  | '03'
  | '04'
  | '05'
  | '06'
  | '07'
  | '08'
  | '09'
  | '10'
  | '11'
  | '12';

type TrendObject = {
  accessibilityLabel: string,
  value: number,
};

type Props = {
  /**
   * A valid color code from the [data visualization palette](https://gestalt.pinterest.systems/foundations/data_visualization/color/palette).
   */
  color?: DataVisualizationColors,
  /**
   * Indicates if TileData should be disabled. Disabled TileDatas are inactive and cannot be interacted with. See the [disabled variant](https://gestalt.pinterest.systems/web/tiledata#Disabled) to learn more.
   */
  disabled?: boolean,
  /**
   * An optional identifier to be passed back in the `onTap` callback. It can be helpful to distinguish multiple TileDatas.
   */
  id?: string,
  /**
   * Handler called when the item selection state is changed.
   */
  onTap?: TileChangeHandler,
  /**
   * Controls whether TileData is selected or not. Use this prop along with the `onTap` handler.
   */
  selected?: boolean,
  /**
   * Shows a visible checkbox when TileData is in a selected state. See the [group variant](https://gestalt.pinterest.systems/web/tiledata#Group) to learn more.
   */
  showCheckbox?: boolean,
  /**
   * The header text for TileData.
   */
  title: string,
  /**
   * Adds a tooltip on hover/focus of TileData. See the [with tooltip](https://gestalt.pinterest.systems/web/tiledata#Tooltip) variant to learn more.
   */
  tooltip?: TooltipProps,
  /**
   * Object detailing the trend value (change in time - e.g. +30%), and accessibility label to describe the trend's icon (e.g., "Trending up").  See the [trend variant](https://gestalt.pinterest.systems/web/datapoint#Trend) to learn more.
   */
  trend?: TrendObject,
  /**
   * A visual indicator whether the trend is considered "good", "bad" or "neutral". By setting \`trendSentiment\` to \`auto\`, a positive trend will be considered "good", a negative trend will be considered "bad" and a trend of zero will be considered "neutral".  See the [trendSentiment variant](https://gestalt.pinterest.systems/web/datapoint#Trend-sentiment) to learn more.
   */
  trendSentiment?: 'good' | 'bad' | 'neutral' | 'auto',
  /**
   * The datapoint value (e.g., 1.23M).
   */
  value: string,
};

/**
 * [TileData](https://gestalt.pinterest.systems/web/tiledata) enables users to select multiple categories to compare with each other in a graph or chart view, while still being able to see all of the data points.
 *
 * ![TileData light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/TileData.spec.mjs-snapshots/TileData-chromium-darwin.png)
 * ![TileData dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/TileData-dark.spec.mjs-snapshots/TileData-dark-chromium-darwin.png)
 *
 */
export default function TileData({
  color = '05',
  disabled = false,
  id,
  onTap,
  selected,
  showCheckbox,
  tooltip,
  title,
  trend,
  trendSentiment,
  value,
}: Props): ReactNode {
  const { colorSchemeName } = useColorScheme();
  const borderColor = getDataVisualizationColor(colorSchemeName, color);

  const colorStyles: { borderColor?: string, backgroundColor?: string } = {
    borderColor,
  };

  const { handleOnBlur, handleOnMouseEnter, handleOnMouseLeave, isHovered } =
    useInteractiveStates();

  const checkboxId = useId();

  const getClasses = () =>
    classnames(styles.baseTile, styles.tileWidth, {
      [styles.hovered]: isHovered,
      [styles.disabled]: disabled,
      [styles[`dataVisualizationColor${color}`]]: selected && !disabled,
    });

  const tileStyle = selected && !disabled ? colorStyles : {};
  const checkBoxStyle = getCheckboxColors({
    state: { hovered: isHovered, selected: !!selected, disabled },
    colorStyles,
  });

  return (
    <MaybeTooltip disabled={disabled} tooltip={tooltip}>
      <Box>
        <TapArea
          disabled={disabled}
          onBlur={handleOnBlur}
          onMouseEnter={handleOnMouseEnter}
          onMouseLeave={handleOnMouseLeave}
          onTap={({ event }) => onTap?.({ event, id, selected: !selected })}
          role="button"
          rounding={4}
        >
          <div className={getClasses()} style={tileStyle}>
            <Flex direction="row" gap={2}>
              <InternalDatapoint
                disabled={disabled}
                lineClamp={2}
                maxTitleWidth={135}
                minTitleWidth={80}
                numTitleRows={2}
                title={title}
                trend={trend}
                trendSentiment={trendSentiment}
                value={value}
              />
              {showCheckbox && (
                <InternalCheckbox
                  checked={selected}
                  id={`readonly-checkbox-blah-${checkboxId}`}
                  readOnly
                  size="sm"
                  style={checkBoxStyle}
                />
              )}
            </Flex>
          </div>
        </TapArea>
      </Box>
    </MaybeTooltip>
  );
}

TileData.displayName = 'TileData';
