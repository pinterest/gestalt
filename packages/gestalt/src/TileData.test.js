// @flow strict
import { create } from 'react-test-renderer';
import Flex from './Flex';
import TileData from './TileData';

describe('TileData', () => {
  it('renders', () => {
    const tree = create(
      <TileData onTap={() => {}} selected title="Text Impressions" value="1.23M" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('TileData renders with tooltip', () => {
    const component = create(
      <TileData
        onTap={() => {}}
        selected
        title="Text Impressions"
        tooltip={{ text: 'This is a tooltip' }}
        value="1.23M"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('TileData renders all colors as expected', () => {
    const colors = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'];
    const component = create(
      <Flex gap={2}>
        {colors.map((color) => (
          <TileData
            key={color}
            color={color}
            onTap={() => {}}
            selected
            title="Text Impressions"
            tooltip={{ text: 'This is a tooltip' }}
            value="1.23M"
          />
        ))}
      </Flex>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Tiledata is disabled', () => {
    const component = create(
      <TileData
        disabled
        onTap={() => {}}
        selected
        showCheckbox
        title="Text Impressions"
        tooltip={{ text: 'This is a tooltip' }}
        value="1.23M"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('TileData title truncates when really long', () => {
    const component = create(
      <TileData
        onTap={() => {}}
        selected
        title="This is some super super super long text and you shouldn't be able to read the last part of it. It's going to use ellipsis to cutoff the last part"
        tooltip={{ text: 'This is a tooltip' }}
        value="1.23M"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('TileData shows readonly checkboxes', () => {
    const component = create(
      <Flex gap={2}>
        <TileData
          onTap={() => {}}
          selected
          showCheckbox
          title=""
          tooltip={{ text: 'This is a tooltip' }}
          value="1.43M"
        />
        <TileData
          onTap={() => {}}
          selected
          showCheckbox
          title=""
          tooltip={{ text: 'This is a tooltip' }}
          value="1.13M"
        />
        <TileData
          onTap={() => {}}
          selected
          showCheckbox
          title=""
          tooltip={{ text: 'This is a tooltip' }}
          value="1.13M"
        />
      </Flex>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
