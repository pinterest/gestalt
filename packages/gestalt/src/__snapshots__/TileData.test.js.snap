// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileData TileData renders with tooltip 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="box xsDisplayBlock"
  >
    <div
      aria-label="this is a tooltip"
      className="box"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        aria-pressed={true}
        className="tile selected focused"
        onBlur={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="button"
        style={
          Object {
            "backgroundColor": "#0081fe1A",
            "borderColor": "#0081fe",
          }
        }
        tabIndex={0}
      >
        <div
          className="Flex rowGap2 columnGap2 xsDirectionRow"
        >
          <div
            className="FlexItem"
          >
            <div
              className="Flex rowGap0 columnGap1 xsDirectionColumn"
            >
              <div
                className="FlexItem"
              >
                <div
                  className="Flex rowGap1 columnGap0 xsDirectionRow xsItemsCenter"
                  style={
                    Object {
                      "minHeight": 24,
                    }
                  }
                >
                  <div
                    className="FlexItem"
                  >
                    <div
                      className="Text fontSize200 defaultText alignStart breakWord fontWeightNormal"
                    >
                      Text Impressions
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="FlexItem"
              >
                <div
                  className="Flex rowGap2 columnGap2 xsDirectionRow xsItemsCenter"
                >
                  <div
                    className="FlexItem"
                  >
                    <div
                      className="Text fontSize400 defaultText alignStart breakWord fontWeightSemiBold"
                    >
                      1.23M
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TileData renders 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    aria-pressed={true}
    className="tile selected focused"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="button"
    style={
      Object {
        "backgroundColor": "#0081fe1A",
        "borderColor": "#0081fe",
      }
    }
    tabIndex={0}
  >
    <div
      className="Flex rowGap2 columnGap2 xsDirectionRow"
    >
      <div
        className="FlexItem"
      >
        <div
          className="Flex rowGap0 columnGap1 xsDirectionColumn"
        >
          <div
            className="FlexItem"
          >
            <div
              className="Flex rowGap1 columnGap0 xsDirectionRow xsItemsCenter"
              style={
                Object {
                  "minHeight": 24,
                }
              }
            >
              <div
                className="FlexItem"
              >
                <div
                  className="Text fontSize200 defaultText alignStart breakWord fontWeightNormal"
                >
                  Text Impressions
                </div>
              </div>
            </div>
          </div>
          <div
            className="FlexItem"
          >
            <div
              className="Flex rowGap2 columnGap2 xsDirectionRow xsItemsCenter"
            >
              <div
                className="FlexItem"
              >
                <div
                  className="Text fontSize400 defaultText alignStart breakWord fontWeightSemiBold"
                >
                  1.23M
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
